app/bundles/LeadBundle/Tests/EventListener/ReportNormalizeSubscriberTest.php

```diff
@@ -0,0 +1,108 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Mautic\LeadBundle\Tests\EventListener;
+
+use Mautic\LeadBundle\Entity\LeadFieldRepository;
+use Mautic\LeadBundle\EventListener\ReportNormalizeSubscriber;
+use Mautic\LeadBundle\Model\FieldModel;
+use Mautic\ReportBundle\Entity\Report;
+use Mautic\ReportBundle\Event\ReportDataEvent;
+use PHPUnit\Framework\TestCase;
+
+class ReportNormalizeSubscriberTest extends TestCase
+{
+    /**
+     * @dataProvider normalizeData
+     *
+     * @param array<int, array<string, array<string, array<string, array<int,string>>|string>|string>> $properties
+     */
+    public function testOnReportDisplay(string $value, string $type, array $properties, string $expected): void
+    {
+        $fieldModel    = $this->createMock(FieldModel::class);
+
+        $fields = [
+            'field1' => [
+                'alias'      => 'field1',
+                'type'       => $type,
+                'properties' => $properties,
+            ],
+        ];
+
+        $leadRepository = $this->createMock(LeadFieldRepository::class);
+        $leadRepository->method('getFields')->willReturn($fields);
+        $fieldModel->method('getRepository')->willReturn($leadRepository);
```

Comment:
Do you think the real query could be used as it's also new? And instead create the field you need for the test?
```suggestion
        $fieldModel = self::$container->get('mautic.lead.model.field');
        \assert($fieldModel instanceof FieldModel::class);
        
        $field = new LeadField();
        $field->setType($type);
        $field->setObject('lead');
        $field->setAlias('field1');
        $field->setName($alias);
        $field->setProperties($properties);

        $fieldModel->save($field);
```

Reply:
Apologies, Zdeno! I should have mentioned that the above was untested code that I just wrote in the review textarea. Please polish it so it passes.

Reply:
I do not understand what you asking for. Do you mean rewrite tests to functional tests?

Reply:
Exactly. So that the new `getFields` method is tested as well.

Reply:
And it already is a functional test. At least the second half.

Reply:
I don't know what you want, sorry. What was wrong with the tests before? What expects after?

Reply:
The idea is that instead of mocking of what fields will be returned from the `getFields` method we simply store the field you need for testing to the database and let the code fetch it from the database. Basically getting closer to what is actually happening in the code when a user is using the app instead of mocking it. The rest of the test should pass as before.